Demo/MozillaSocial-iOS/MoSoContent/Auth.swift

```diff
@@ -0,0 +1,138 @@
+
+    var authToken: Token?
+
+    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
+        return ASPresentationAnchor()
+    }
+
+    func registerApp() async -> ClientEntity? {
+        var urlRequest = URLRequest(url: registerAppURL())
+        urlRequest.httpMethod = "POST"
+
+        do{
+            let (data, _) = try await URLSession.shared.data(for: urlRequest)
+
+            let decoder = JSONDecoder()
+            decoder.keyDecodingStrategy = .convertFromSnakeCase
+            return try decoder.decode(ClientEntity.self, from: data)
+        } catch {
+            print("error: ", error)
+            return nil
+        }
+    }
+
+    func signIn(client: ClientEntity) {
+        var authURL = signInURL(with: client.clientId)
+
+        let session = ASWebAuthenticationSession(url: authURL, callbackURLScheme: redirectScheme) { (url, error) in
+            guard error == nil else {
+                print("ERROR: \(String(describing: error))")
+                return
+            }
+
+            guard let url = url, let code = self.parseCode(from: url) else {
+                print("Invalid Code Return")
+                return
+            }
+
+            Task {
+                self.authToken = await self.fetchAuthToken(for: client, with: code)
+            }
+        }
+
+        session.presentationContextProvider = self
+        session.prefersEphemeralWebBrowserSession = true
+        session.start()
+    }
+
+    private func fetchAuthToken(for client: ClientEntity, with code: String) async -> Token? {
+        var urlRequest = URLRequest(url: authTokenURL(for: client, with: code))
+        urlRequest.httpMethod = "POST"
+        do{
+            let (data, _) = try await URLSession.shared.data(for: urlRequest)
+
+            let decoder = JSONDecoder()
+            decoder.keyDecodingStrategy = .convertFromSnakeCase
+
+            return try decoder.decode(Token.self, from: data)
+        } catch {
+            print("error: ", error)
+        }
+
+        return nil
+    }
+
+    // MARK: - URL Builders
+
+    private func authTokenURL(for client: ClientEntity, with code: String) -> URL {
+        guard var url = URL(string: baseURL + "/oauth/token") else {
+            fatalError()
+        }
+
+        url.append(queryItems:[URLQueryItem(name: "code", value: code),
+                               URLQueryItem(name: "grant_type", value: "authorization_code"),
+                               URLQueryItem(name: "redirect_uri", value: redirectURI),
+                               URLQueryItem(name: "client_secret", value: client.clientSecret),
+                               URLQueryItem(name: "client_id", value: client.clientId),
+                               URLQueryItem(name: "scope", value: "read+write+push+follow")])
+        return url
+    }
+
+    private func registerAppURL() -> URL {
+        guard var registerURL = URL(string: baseURL + "/api/v1/apps") else {
```

Comment:
can we extract `"/api/v1/apps"` into a constant, and put it in the same `enum` mentioned in a previous review comment? Also, we could add a `static func` to that enum that returns the entire url string, so you don't have to do the calculation at the caller site.

Reply:
Since the URLs are not used in any other places I thought it fine to leave them in place.